### DISCUSSION TITLE
[Unreal]call TsConstruct of的报错

### DISCUSSION THREAD

**Discussion by watsonsong** [Tencent/puerts#1439] [Q&A]
工程在某次更新后，某个TS引擎类继承的GameModeBase，在打包的Windows版本会出现如下报错。但在编辑器中是正常的：
```call TsConstruct of XXX_GameMode_C_2147482227(0000026FD382C480) fail!, DynamicInvoker invalid```

之前这个TS类有自定义的Constructor，后来删除掉了。也强制重新生成过蓝图代理。尝试关闭掉AsyncLoadingThreadEnabled，都没有效果。目前没有太好的思路来排查，想请教一下还有哪些方面可以着手检查一下？

**Comment by watsonsong** [comment]
发现产生如下报错的原因是因为JS的初始化顺序。
···
Puerts: Error: call TsConstruct of BP_Dungeon_GameMode_C_2147482217(000001C215A4F200) fail!, DynamicInvoker invalid
Puerts: Error: call TsConstruct of DungeonGameState_C_2147482206(000001C215A48900) fail!, DynamicInvoker invalid
Puerts: Error: call TsConstruct of DungeonPlayerState_C_2147482201(000001C1C026CCD0) fail!, DynamicInvoker invalid
Puerts: Error: call DungeonGameMode_C::GetDefaultPawnClassForController fail!, DynamicInvoker invalid
···

当我把Puerts.upluginz中的Puerts的LoadingPhase改为PreDefault（顺便PuertsEditor也改回PuertsEditor），问题就解决了。

我发现编辑器在启动的时候，这个GameMode就因为某种原因是最开始加载的，PostEngineInit会比它晚。这种情况下JsEnv的Rebind逻辑似乎并没有生效。

**Reply by 8197492** [reply]
多谢大哥，改 了PreDefault，果然就好了。

**Reply by watsonsong** [reply]
但是这个现象很奇怪，没找到根本原因。按道理使用PostEngineInit也不会有问题才对。虚拟机启动的时候会遍历所有uobject去做绑定的，但实际情况这个过程似乎失效了。改成PreDefault只是一定程度上避免了虚拟机启动前有蓝图代理独享提前加载的情况而已。
